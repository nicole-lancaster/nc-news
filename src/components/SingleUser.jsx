import { useState } from "react";

const SingleUser = ({ user, currentUser, setCurrentUser }) => {
  const [isLoading, setIsLoading] = useState(false);

  const handleLoginClick = (user) => {
    setIsLoading(true);
    setCurrentUser(user);
    setIsLoading(false);
  };

  if (isLoading) return <p>Logging in...</p>;

  return (
    <section className="single-user ">
      <button
        className="user-pretend-to-be-btn border border-white bg-none hover:border-black"
        onClick={() => handleLoginClick(user)}
        disabled={user.username === currentUser?.username}
      >
        <li className="single-user-name">
          {user.name}
          <p className="single-user-username">{user.username}</p>
          <img
            className="user-list-avatar"
            src={user.avatar_url}
            alt={user.username}
          />
        </li>
      </button>
      {currentUser === user ? (
        <p className="user-logged-in-as-msg">
          You are now logged in as {user.name}!
        </p>
      ) : null}
    </section>
  );
};
export default SingleUser;
