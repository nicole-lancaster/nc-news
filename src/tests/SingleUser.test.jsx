import { render, screen, waitFor } from "@testing-library/react";
import { UserContext } from "../contexts/User";
import SingleUser from "../components/SingleUser";
import "@testing-library/jest-dom";

jest.mock("../api.js", () => ({
  fetchUsers: () =>
    Promise.resolve([
      { name: "Rob", username: "robR" },
      { name: "Nicole", username: "nicoleL" },
    ]),
}));

describe("SingleUser component", () => {
  test("should match rendered snapshot", () => {
    const user = {
      name: "Nicole",
      username: "nicoleL",
      avatar_url: "path/to/img",
    };
    const { asFragment } = render(
      <UserContext.Provider
        value={{ currentUser: null, setCurrentUser: () => {} }}
      >
        <SingleUser user={user} />
      </UserContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("should display single user", async () => {
    const user = {
      name: "Nicole",
      username: "nicoleL",
      avatar_url: "path/to/img",
    };
    render(
      <UserContext.Provider
        value={{ currentUser: null, setCurrentUser: () => {} }}
      >
        <SingleUser user={user} />
      </UserContext.Provider>
    );
    await waitFor(() => {
      expect(screen.getByText("Nicole as nicoleL")).toBeInTheDocument();
    });
    await waitFor(() => {
      expect(screen.getByAltText("nicoleL")).toBeInTheDocument();
    });
    await waitFor(() => {
      expect(screen.getByAltText("nicoleL")).toHaveAttribute(
        "src",
        "path/to/img"
      );
    });
  });
  test("should disable login button when user is already logged in", () => {
    const user = { name: "Rob", username: "robR", avatar_url: "path/to/img" };
    render(
      <UserContext.Provider
        value={{ currentUser: user, setCurrentUser: () => {} }}
      >
        <SingleUser user={user} />
      </UserContext.Provider>
    );
    expect(screen.getByTestId("login-button")).toBeDisabled();
  });
});
