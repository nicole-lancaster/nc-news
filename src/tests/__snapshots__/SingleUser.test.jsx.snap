// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleUser component should match rendered snapshot 1`] = `
<DocumentFragment>
  <section>
    <div
      class="border-2 border-light-gray-700 shadow rounded-lg hover:border-pink-700 m-1 p-1 w-32 md:w-48 lg:m-2 xl:w-120"
    >
      <button
        data-testid="login-button"
        type="submit"
      >
        <li
          class="flex flex-col items-center mx-1 my-2 "
        >
          <p
            class="text-center text-xs md:text-sm lg:text-base xl:text-lg font-mono mt-2"
          >
            Nicole as nicoleL
          </p>
          <img
            alt="nicoleL"
            class="p-1 w-fit h-20 self-center"
            src="path/to/img"
          />
        </li>
      </button>
    </div>
  </section>
</DocumentFragment>
`;
